frame/vesting/src/vesting_info.rs

```diff
@@ -0,0 +1,131 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Module to enforce private fields on `VestingInfo`.
+
+use super::*;
+use codec::MaxEncodedLen;
+
+/// Struct to encode the vesting schedule of an individual account.
+#[derive(Encode, Decode, Copy, Clone, PartialEq, Eq, RuntimeDebug, MaxEncodedLen)]
+pub struct VestingInfo<Balance, BlockNumber> {
+	/// Locked amount at genesis.
+	locked: Balance,
+	/// Amount that gets unlocked every block after `starting_block`.
+	per_block: Balance,
+	/// Starting block for unlocking(vesting).
+	starting_block: BlockNumber,
+}
+
+impl<Balance, BlockNumber> VestingInfo<Balance, BlockNumber>
+where
+	Balance: AtLeast32BitUnsigned + Copy,
+	BlockNumber: AtLeast32BitUnsigned + Copy + Bounded,
+{
+	/// Instantiate a new `VestingInfo`.
+	pub fn new<T: Config>(
+		locked: Balance,
+		per_block: Balance,
+		starting_block: BlockNumber,
+	) -> VestingInfo<Balance, BlockNumber> {
+		VestingInfo { locked, per_block, starting_block }
+	}
+
+	/// Validate parameters for `VestingInfo`. Note that this does not check
+	/// against `MinVestedTransfer` or the current block.
+	pub fn validate<BlockNumberToBalance: Convert<BlockNumber, Balance>, T: Config>(
+		&self,
+	) -> Result<(), Error<T>> {
+		ensure!(!self.locked.is_zero(), Error::<T>::InvalidScheduleParams);
+
+		let max_block = BlockNumberToBalance::convert(BlockNumber::max_value());
+		let starting_block = BlockNumberToBalance::convert(self.starting_block());
+		match self.locked.checked_div(&self.per_block) {
+			None => return Err(Error::<T>::InfiniteSchedule), // `per_block` is 0
+			Some(duration) => {
+				let end = duration.saturating_add(starting_block);
+				ensure!(end < max_block, Error::<T>::InfiniteSchedule)
+			}
+		};
+
+		Ok(())
+	}
+
+	/// Potentially correct the `per_block` of a schedule. Typically called after `validate`.
+	pub fn correct(mut self) -> Self {
+		self.per_block = if self.per_block > self.locked { self.locked } else { self.per_block };
+		self
+	}
+
+	/// Locked amount at schedule creation.
+	pub fn locked(&self) -> Balance {
+		self.locked
+	}
+
+	/// Amount that gets unlocked every block after `starting_block`.
+	pub fn per_block(&self) -> Balance {
+		self.per_block
+	}
+
+	/// Starting block for unlocking(vesting).
+	pub fn starting_block(&self) -> BlockNumber {
+		self.starting_block
+	}
+
+	/// Amount locked at block `n`.
+	pub fn locked_at<BlockNumberToBalance: Convert<BlockNumber, Balance>>(
+		&self,
+		n: BlockNumber,
+	) -> Balance {
+		// Number of blocks that count toward vesting
+		// Saturating to 0 when n < starting_block
+		let vested_block_count = n.saturating_sub(self.starting_block);
+		let vested_block_count = BlockNumberToBalance::convert(vested_block_count);
+		// Return amount that is still locked in vesting
+		vested_block_count
+			.checked_mul(&self.per_block)
```

Comment:
```suggestion
			.checked_mul(&self.per_block.max(One::one()))
```

This solves any existing "bad" vesting schedules right?

Reply:
Yeah, we should decide though what we want to do in general with pre existing "bad" schedules. Do we want to try and correct them or should we error?

For example here we error:  https://github.com/paritytech/substrate/pull/9202/files/48dd1b5fafeede2384acb96d5f1d83ea1eae8273#diff-404daa3dacc566f4515f7058e049c1cfe33bd69694ae247215da9c9a9c1c6752R117

Reply:
why would we need to write anything to storage if we have the logic suggested here?

In the calculation phase, we take any 0 and make it a 1, so no need to update or backfill anyone.

Reply:
ee6a75dcb9deb7262dab66f74270a67dcb8b6db5